src/planner/partialize.c

```diff
@@ -185,6 +197,656 @@ partialize_agg_paths(RelOptInfo *rel)
+	partially_grouped_rel->reloptkind = input_rel->reloptkind;
+	partially_grouped_rel->serverid = input_rel->serverid;
+	partially_grouped_rel->userid = input_rel->userid;
+	partially_grouped_rel->useridiscurrent = input_rel->useridiscurrent;
+	partially_grouped_rel->fdwroutine = input_rel->fdwroutine;
+
+	/* Build target list for partial aggregate paths */
+	PathTarget *grouping_target = output_rel->reltarget;
+	PathTarget *partial_grouping_target = ts_make_partial_grouping_target(root, grouping_target);
+	partially_grouped_rel->reltarget = partial_grouping_target;
+
+	/* Calculate aggregation costs */
+	if (!extra_data->partial_costs_set)
+	{
+		/* partial phase */
+		get_agg_clause_costs_compat(root,
+									(Node *) partial_grouping_target->exprs,
+									AGGSPLIT_INITIAL_SERIAL,
+									&extra_data->agg_partial_costs);
+
+#if PG14_LT
+		PathTarget *target = root->upper_targets[UPPERREL_GROUP_AGG];
+#endif
+
+		/* final phase */
+		get_agg_clause_costs_compat(root,
+									(Node *) target->exprs,
+									AGGSPLIT_FINAL_DESERIAL,
+									&extra_data->agg_final_costs);
+
+		extra_data->partial_costs_set = true;
+	}
+
+	/* Generate the aggregation pushdown path */
+	Path *cheapest_total_path = get_best_total_path(input_rel);
+	Assert(cheapest_total_path != NULL);
+	generate_agg_pushdown_path(root,
+							   cheapest_total_path,
+							   output_rel,
+							   partially_grouped_rel,
+							   grouping_target,
+							   partial_grouping_target,
+							   can_sort,
+							   can_hash,
+							   d_num_groups,
+							   extra_data);
+
+	/* The same as above but for partial paths */
+	if (input_rel->partial_pathlist != NIL && input_rel->consider_parallel)
+	{
+		Path *cheapest_partial_path = linitial(input_rel->partial_pathlist);
+		generate_partial_agg_pushdown_path(root,
+										   cheapest_partial_path,
+										   output_rel,
+										   partially_grouped_rel,
+										   grouping_target,
+										   partial_grouping_target,
+										   can_sort,
+										   can_hash,
+										   d_num_groups,
+										   extra_data);
+	}
+
+	/* Replan aggregation if we were able to generate partially grouped rel paths */
+	if (partially_grouped_rel->pathlist == NIL)
+		return;
+
+	/* Prefer our paths */
+	output_rel->pathlist = NIL;
+	output_rel->partial_pathlist = NIL;
+
+	/* Finalize partially aggregated append paths */
+	AggClauseCosts *agg_final_costs = &extra_data->agg_final_costs;
+	ListCell *lc;
+	foreach (lc, partially_grouped_rel->pathlist)
+	{
+		Path *append_path = lfirst(lc);
+
+		if (is_plain_or_sorted_agg_path(append_path))
+		{
+			bool is_sorted;
+			int presorted_keys;
+
+			is_sorted = pathkeys_count_contained_in(root->group_pathkeys,
+													append_path->pathkeys,
+													&presorted_keys);
+
+			if (!is_sorted)
+			{
+				append_path = (Path *)
+					create_sort_path(root, output_rel, append_path, root->group_pathkeys, -1.0);
+			}
```

Comment:
I have a few questions about  this:
1) So we are sorting the results of append of partial aggregation before the "finalize aggregate" node? Why is it required?
2) Anyway, the partial aggregates should require the same sorting, so putting a merge append node over them would avoid the need for sorting?
3) Do we need aggregation over explicit sort at all? In terms of algorithmic complexity, and in most practical cases as well, hash aggregation is preferrable if the input data is not sorted.

Reply:
1) It is required because finalizing a sorted aggregation requires a sorted input of partials. The aggregation node assumes that an aggregation group is finished after a new group key is seen. So, the node would emit the same aggregation group multiple times if the input is unordered.
2) Yes, this could be used to prevent the re-sorting. However, this would replace our `ChunkAppend` node. This PR allows it to use Append/MergeAppend/ChunkAppend nodes between the `AGGSPLIT_INITIAL_SERIAL` and the `AGGSPLIT_FINAL_DESERIAL` aggregation node.
3) I agree. In most cases, the hash aggregation is preferable. However, I create both paths if possible (_hashed_ and _sorted/plain_) and let PostgreSQL decide which one to use. 

Note: 1 & 3 follow the same pattern as the PostgreSQL [planner code](https://github.com/postgres/postgres/blob/e1c6db6309292827fc6baf7cf7df87d49254c7e1/src/backend/optimizer/plan/planner.c#L7101) for partitioned tables.